File: apps/windows/TeslaSync.App/shared-surfaces/RouteTransition/RouteTransition.Model.cs
namespace TeslaSync.App.SharedSurfaces;

/// <summary>
/// Canonical metadata for the <c>RouteTransition</c> shared surface — the native mirror of the web component at
/// <c>web/src/components/motion/RouteTransition.tsx</c>. The web source wraps the routed page body (the router
/// <c>&lt;Outlet /&gt;</c>) and cross-fades it on every <c>location.pathname</c> change: a 120ms ease-out fade
/// plus a 4px vertical translate, with <c>mode="wait"</c> so the outgoing page leaves before the incoming page
/// enters. It re-keys by pathname only (query/search/hash changes never re-fade), skips the very first mount
/// (<c>initial={false}</c>), collapses to a no-op under <c>prefers-reduced-motion</c>, and suppresses the fade
/// entirely for list↔detail navigations (drilling into <c>/drives/:id</c> and back) so those feel instant. The
/// surface is anonymous: it renders no visible chrome and no static copy, so there are no i18n keys to resolve
/// and no interactive elements. This holder pins the diagnostics slug, the automation id, the default fade
/// duration and translate offset, the accessibility-transparency contract and the default skip-pattern list.
/// UI-free so the metadata is asserted headlessly without a XAML runtime.
/// </summary>
public static class RouteTransitionRegistration
{
    /// <summary>Diagnostics surface slug emitted with the <c>view.opened</c> event (P1/S11 contract).</summary>
    public const string Slug = "RouteTransition";

    /// <summary>
    /// The root automation id the view stamps on itself. The web component is an anonymous <c>motion.div</c>
    /// wrapper with no <c>data-testid</c>, so this is the native-only stable hook UI-automation tests target.
    /// </summary>
    public const string RootAutomationId = "route-transition";

    /// <summary>
    /// Default cross-fade duration in milliseconds (web <c>useMotionPreference(120)</c>). The fade is subtle
    /// enough to feel polished without slowing the user down.
    /// </summary>
    public const int DefaultDurationMs = 120;

    /// <summary>
    /// Default vertical translate magnitude in pixels (web <c>y: 4</c> on enter / <c>y: -4</c> on exit). The
    /// incoming page rises into place and the outgoing page lifts away.
    /// </summary>
    public const double DefaultOffsetY = 4.0;

    /// <summary>
    /// False because the wrapper introduces no semantic node of its own (the web <c>motion.div</c> carries no
    /// ARIA role). The view keeps itself out of the Narrator control view and out of the tab order so the routed
    /// page body it hosts is announced directly, with no extra container node to step through.
    /// </summary>
    public const bool ContributesAccessibilityNode = false;

    /// <summary>
    /// The route patterns whose navigations are NOT animated (web <c>DEFAULT_SKIP_PATTERNS</c>). Drilling from a
    /// list (<c>/drives</c>) into a detail (<c>/drives/123</c>) — and back out — feels better when it is
    /// near-instant, so the cross-fade is suppressed when EITHER the previous or the new pathname matches any of
    /// these (react-router <c>matchPath</c> with <c>end: true</c>). Order does not matter — the first match wins.
    /// </summary>
    public static IReadOnlyList<string> DefaultSkipPatterns { get; } = Array.AsReadOnly(new[]
    {
        "/drives/:id",
        "/drives/:id/replay",
        "/charging/:id",
        "/vehicles/:id",
        "/vehicles/:id/access",
        "/trips/:id",
    });
}

/// <summary>
/// Pure route-pattern matcher — the native port of react-router's <c>matchPath({ path, end: true }, pathname)</c>
/// the web source calls for each skip pattern (web/src/components/motion/RouteTransition.tsx). A pattern is split
/// into <c>/</c>-separated segments; a static segment must equal the corresponding path segment
/// case-insensitively (react-router's default <c>caseSensitive: false</c>), and a <c>:param</c> segment matches
/// exactly one non-empty path segment (react-router compiles <c>/:id</c> to <c>/([^/]+)</c>). With
/// <c>end: true</c> the whole path must be consumed, so the segment counts must match; a trailing slash on either
/// side is tolerated (react-router appends <c>\/*$</c> when matching to the end). No WinUI types and no regex
/// dependency — unit-tested without a UI host.
/// </summary>
public static class RouteMatcher
{
    /// <summary>
    /// Whether <paramref name="pathname"/> matches the react-router <paramref name="pattern"/> end-to-end.
    /// </summary>
    /// <param name="pattern">A react-router path pattern (e.g. <c>/drives/:id</c>).</param>
    /// <param name="pathname">The current location path (web <c>location.pathname</c>).</param>
    public static bool Matches(string pattern, string pathname)
    {
        ArgumentNullException.ThrowIfNull(pattern);
        ArgumentNullException.ThrowIfNull(pathname);

        string[] patternSegments = Split(pattern);
        string[] pathSegments = Split(pathname);

        // end: true — the full path must be consumed, so the segment counts have to line up exactly.
        if (patternSegments.Length != pathSegments.Length)
        {
            return false;
        }

        for (int i = 0; i < patternSegments.Length; i++)
        {
            string patternSegment = patternSegments[i];

            if (patternSegment.StartsWith(':'))
            {
                // react-router /:param → /([^/]+): a dynamic segment matches exactly one non-empty segment.
                if (pathSegments[i].Length == 0)
                {
                    return false;
                }

                continue;
            }

            if (!string.Equals(patternSegment, pathSegments[i], StringComparison.OrdinalIgnoreCase))
            {
                return false;
            }
        }

        return true;
    }

    /// <summary>
    /// Whether <paramref name="pathname"/> matches ANY of <paramref name="patterns"/> (web
    /// <c>skipPattern.some(...)</c>).
    /// </summary>
    /// <param name="patterns">The route patterns to test against.</param>
    /// <param name="pathname">The current location path.</param>
    public static bool MatchesAny(IReadOnlyList<string> patterns, string pathname)
    {
        ArgumentNullException.ThrowIfNull(patterns);
        ArgumentNullException.ThrowIfNull(pathname);

        for (int i = 0; i < patterns.Count; i++)
        {
            if (Matches(patterns[i], pathname))
            {
                return true;
            }
        }

        return false;
    }

    private static string[] Split(string path)
    {
        // Normalise the leading slash and ignore a trailing one (react-router's `^\/*` / `\/*$` handling); the
        // root path "/" yields zero segments.
        string trimmed = path.Trim('/');
        return trimmed.Length == 0 ? Array.Empty<string>() : trimmed.Split('/');
    }
}

/// <summary>
/// Pure projection of one navigation into the transition the view plays — the native port of the web component
/// body (web/src/components/motion/RouteTransition.tsx lines computing <c>skipForList</c> and
/// <c>effectiveDurationMs</c>). It decides whether the cross-fade runs (<see cref="Animate"/>), the effective
/// fade <see cref="DurationMs"/> and the vertical <see cref="OffsetY"/>, and records WHY a fade was suppressed
/// (<see cref="Reduced"/> / <see cref="SkippedForList"/>). The fade is suppressed when the user prefers reduced
/// motion OR when either the previous or the new pathname matches a skip pattern (a list↔detail navigation); in
/// both cases the duration and offset collapse to zero so the swap is instant — exactly the web
/// <c>reduce || skipForList ? 0 : durationMs</c> and the <c>initial/exit</c> branches that hold opacity and y
/// fixed. Kept static and side-effect-free so the adapter is unit-testable without a view-model or a UI thread.
/// </summary>
public readonly record struct RouteTransitionPlan
{
    private RouteTransitionPlan(bool animate, int durationMs, double offsetY, bool reduced, bool skippedForList)
    {
        Animate = animate;
        DurationMs = durationMs;
        OffsetY = offsetY;
        Reduced = reduced;
        SkippedForList = skippedForList;
    }

    /// <summary>
    /// Whether the cross-fade runs. False under reduced motion, a list↔detail skip, or a non-positive duration,
    /// where the page swap is instant (web <c>effectiveDurationMs === 0</c>).
    /// </summary>
    public bool Animate { get; }

    /// <summary>
    /// The effective fade duration in milliseconds (web <c>effectiveDurationMs</c>): the configured duration when
    /// animating, otherwise <c>0</c>.
    /// </summary>
    public int DurationMs { get; }

    /// <summary>
    /// The effective vertical translate magnitude in pixels (web <c>y: 4</c> / <c>-4</c>): the configured offset
    /// when animating, otherwise <c>0</c> (the web <c>initial/exit</c> states hold <c>y</c> at 0 when suppressed).
    /// </summary>
    public double OffsetY { get; }

    /// <summary>True when the fade was suppressed because the user prefers reduced motion.</summary>
    public bool Reduced { get; }

    /// <summary>
    /// True when the fade was suppressed because the previous or new pathname matched a skip pattern (web
    /// <c>skipForList</c>) — a list↔detail drill-in / drill-out.
    /// </summary>
    public bool SkippedForList { get; }

    /// <summary>
    /// Project a navigation from <paramref name="previousPath"/> to <paramref name="newPath"/> into its
    /// transition. <paramref name="durationMs"/> and <paramref name="offsetY"/> are clamped to be non-negative;
    /// the fade is suppressed when <paramref name="reduceMotion"/> is set or either path matches a
    /// <paramref name="skipPatterns"/> entry, mirroring the web <c>reduce || skipForList</c> guard.
    /// </summary>
    /// <param name="previousPath">The pathname being navigated away from (web <c>prevPathRef.current</c>).</param>
    /// <param name="newPath">The pathname being navigated to (web <c>location.pathname</c>).</param>
    /// <param name="reduceMotion">Whether the OS reduce-motion preference is set (web <c>reduce</c>).</param>
    /// <param name="durationMs">The configured fade duration in milliseconds (web <c>durationMs</c>).</param>
    /// <param name="skipPatterns">The list↔detail patterns that suppress the fade (web <c>skipPattern</c>).</param>
    /// <param name="offsetY">The configured vertical translate magnitude in pixels (web <c>y</c>).</param>
    public static RouteTransitionPlan Compute(
        string previousPath,
        string newPath,
        bool reduceMotion,
        int durationMs,
        IReadOnlyList<string> skipPatterns,
        double offsetY)
    {
        ArgumentNullException.ThrowIfNull(previousPath);
        ArgumentNullException.ThrowIfNull(newPath);
        ArgumentNullException.ThrowIfNull(skipPatterns);

        // web: matchesSkip(prevPath) || matchesSkip(newPath) — back-navigation (POP) is skipped too.
        bool skippedForList =
            RouteMatcher.MatchesAny(skipPatterns, previousPath) ||
            RouteMatcher.MatchesAny(skipPatterns, newPath);

        int safeDuration = durationMs < 0 ? 0 : durationMs;
        double safeOffset = offsetY < 0 ? 0 : offsetY;
        bool suppressed = reduceMotion || skippedForList;
        bool animate = !suppressed && safeDuration > 0;

        return new RouteTransitionPlan(
            animate,
            animate ? safeDuration : 0,
            animate ? safeOffset : 0,
            reduceMotion,
            skippedForList);
    }
}

/// <summary>
/// PII-safe diagnostics for the <c>RouteTransition</c> surface (P1/S11 diagnostics contract). Records only
/// operational counters with the surface slug — never the route path, which can carry fleet identifiers (a
/// <c>/charging/{id}</c> or <c>/vehicles/{id}</c> path) — so a diagnostics line can never leak where a user
/// navigates. Thread-safe.
/// </summary>
public sealed class RouteTransitionDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;
    private long _animated;
    private long _skipped;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    /// <param name="sink">An optional operational-only line sink (no route path is ever passed).</param>
    public RouteTransitionDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been mounted/opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Number of navigations that played the cross-fade (count only — never the path).</summary>
    public long Animated => Interlocked.Read(ref _animated);

    /// <summary>Number of navigations swapped instantly (reduced motion or a list↔detail skip).</summary>
    public long Skipped => Interlocked.Read(ref _skipped);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=RouteTransition</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={RouteTransitionRegistration.Slug}");
    }

    /// <summary>Record an animated navigation, emitting <c>route.transition slug=RouteTransition</c> (no path).</summary>
    public void RecordAnimated()
    {
        Interlocked.Increment(ref _animated);
        _sink?.Invoke($"route.transition slug={RouteTransitionRegistration.Slug}");
    }

    /// <summary>Record an instant (suppressed) navigation, emitting <c>route.skipped slug=RouteTransition</c>.</summary>
    public void RecordSkipped()
    {
        Interlocked.Increment(ref _skipped);
        _sink?.Invoke($"route.skipped slug={RouteTransitionRegistration.Slug}");
    }
}

File: apps/android/app/src/main/java/com/teslasync/shared-surfaces/RouteTransition/RouteTransitionModel.kt
// Pure, framework-free model + transition projection + diagnostics for the RouteTransition shared surface — the
// native analogue of every decision the web component makes (web/src/components/motion/RouteTransition.tsx)
// before Compose paints a frame. No Compose, no Android, no HTTP: every declaration here runs off-device in the
// :android:testReleaseUnitTest gate, keeping the composable a thin render layer.
//
// What the web source is (and therefore the COMPLETE behaviour this surface reproduces): a motion wrapper placed
// around the routed page body (`<Outlet />`) that cross-fades the content whenever the location pathname changes
// (a 120 ms ease-out fade + 4 px y-translate, `mode="wait"` so the outgoing page unmounts before the incoming
// mounts). The two inputs it reads are `useLocation()` (it re-keys by the pathname ONLY — query/search/hash
// changes such as filters, sort or anchors never re-fade) and `useMotionPreference(120)`. It collapses the fade
// to an instant swap in two cases: when the user has requested reduced motion, and when EITHER the previous or
// the next pathname matches a list↔detail skip pattern (`/drives/:id`, `/charging/:id`, …) so a drill-in and the
// matching drill-back-out both feel snappy. `initial={false}` means the very first render never animates.
//
// Why the generic data-surface states (loading / empty / error / stale / offline) are intentionally absent: this
// surface performs no query. Its only inputs are the current location key (the nav layer, P1/S8) and the device
// motion preference (P1/S8, bound at the render boundary by the motion atom's reduced-motion plumbing) — both
// always available, neither ever "loads", "errors", goes "stale" or goes "offline". It wraps arbitrary page
// content and is transparent to it; the wrapped page owns its own data states and renders them inside this
// container. Inventing those states here would model an async dependency the web spec does not have (honesty
// covenant: no scope narrowing, no silent drift). The surface's REAL, fully-reproduced states are the transition
// branches the web file plays — the animated cross-fade (a non-skip page-to-page navigation with motion
// enabled), and the instant swap (reduced motion, OR a list↔detail navigation in either direction) — each
// projected by [transitionPlan] and asserted off-device so every branch doubles as a per-state snapshot.
//
// The skip-pattern matching + the reduced/skip duration collapse are composed from the component-library motion
// atom's tested primitives (ADR-005: shouldSkipTransition + effectiveDurationMs) rather than re-derived, so the
// surface and the atom can never drift on what "skip a list↔detail fade" means; this model only projects those
// primitives into the full per-navigation plan callers need to know whether a navigation animates.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/shared-surfaces/RouteTransition — the P3 prompt's allowed-files path) cannot form a valid Kotlin
// package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package intentionally
// diverges from the path — exactly as the sibling StaggerContainer / ScrollRestoration surfaces do.
// `MatchingDeclarationName` is suppressed for the co-located supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.sharedsurfaces.routetransition

import io.teslasync.android.components.motion.DEFAULT_SKIP_ROUTE_PATTERNS
import io.teslasync.android.components.motion.MotionDefaults
import io.teslasync.android.components.motion.effectiveDurationMs
import io.teslasync.android.components.motion.shouldSkipTransition
import io.teslasync.shared.core.diagnostics.Logger

/**
 * Diagnostics surface slug emitted with the `view.opened` event (P1/S11). Carries no route, no page content and
 * no user data — only this constant identifier — so a diagnostics line can never leak which screen was shown.
 */
const val ROUTE_TRANSITION_SLUG: String = "RouteTransition"

/**
 * Canonical registry metadata for the RouteTransition surface. The diagnostics [SLUG] is emitted with the
 * one-shot `view.opened` event (P1/S11) and is the surface slug the prompt mandates (`RouteTransition`).
 */
object RouteTransitionRegistration {
    /** Stable surface id. */
    const val ID: String = "route-transition"

    /** Diagnostics surface slug emitted with the `view.opened` event (P1/S11). */
    const val SLUG: String = ROUTE_TRANSITION_SLUG
}

/**
 * Base page-transition length — the native mirror of the web `useMotionPreference(120)` 120 ms cross-fade.
 * Sourced from the motion-token atom ([MotionDefaults.TRANSITION_MS], the design-system `fast` duration) so the
 * surface and the atom share one source of truth for how long a page cross-fade runs.
 */
const val DEFAULT_TRANSITION_DURATION_MS: Int = MotionDefaults.TRANSITION_MS

/**
 * The route patterns whose list↔detail navigations skip the cross-fade — re-exported from the motion atom's
 * [DEFAULT_SKIP_ROUTE_PATTERNS] (the native mirror of the web `DEFAULT_SKIP_PATTERNS`) so callers can read the
 * default set without reaching into the atom package, and so there is a single source of truth for it.
 */
val DEFAULT_ROUTE_SKIP_PATTERNS: List<String> = DEFAULT_SKIP_ROUTE_PATTERNS

/**
 * The resolved render decision for a single navigation from [previousRoute] to a new route — the native analogue
 * of the web component's per-render computation (`skipForList = matchesSkip(prev) || matchesSkip(next)` then
 * `effectiveDurationMs = reduce || skipForList ? 0 : durationMs`). Pure data so the decision is asserted
 * off-device and the composable stays a thin render layer over it.
 *
 * @property effectiveDurationMs how long the cross-fade runs for this navigation; 0 means an instant swap.
 * @property skippedForListDetail true when this is a list↔detail navigation (drill-in or back-out) that suppresses
 *   the fade regardless of the motion preference.
 * @property reduce the active reduced-motion preference that fed the decision.
 */
data class RouteTransitionPlan(
    val effectiveDurationMs: Int,
    val skippedForListDetail: Boolean,
    val reduce: Boolean,
) {
    /** True when the navigation plays a visible cross-fade (i.e. the effective duration is non-zero). */
    val animates: Boolean
        get() = effectiveDurationMs > 0

    /** True when the navigation is an instant swap — reduced motion, a skipped list↔detail hop, or both. */
    val instant: Boolean
        get() = effectiveDurationMs == 0
}

/**
 * Project a navigation from [previousRoute] to [nextRoute] into the render-ready [RouteTransitionPlan]. The fade
 * is suppressed (duration 0) when [reduce] is set OR when either path matches a [skipPatterns] entry — the native
 * mirror of the web `reduce || skipForList ? 0 : durationMs`. A 1:1 port that composes the motion atom's tested
 * [shouldSkipTransition] + [effectiveDurationMs] so the surface and the atom can never disagree on what animates.
 *
 * @param previousRoute the pathname the user is leaving (the web `prevPathRef.current`).
 * @param nextRoute the pathname being navigated to (the web `location.pathname`).
 * @param reduce the active reduced-motion preference (the web `useMotionPreference().reduce`).
 * @param durationMs the base cross-fade length when the navigation does animate (web `durationMs`, default 120 ms).
 * @param skipPatterns the list↔detail patterns that suppress the fade in either direction.
 */
fun transitionPlan(
    previousRoute: String,
    nextRoute: String,
    reduce: Boolean,
    durationMs: Int = DEFAULT_TRANSITION_DURATION_MS,
    skipPatterns: List<String> = DEFAULT_ROUTE_SKIP_PATTERNS,
): RouteTransitionPlan {
    val skipForList = shouldSkipTransition(previousRoute, nextRoute, skipPatterns)
    val effective = effectiveDurationMs(reduce = reduce || skipForList, requestedMs = durationMs)
    return RouteTransitionPlan(
        effectiveDurationMs = effective,
        skippedForListDetail = skipForList,
        reduce = reduce,
    )
}

/**
 * The cross-fade key for a location — the native analogue of the web `key={location.pathname}`. Re-keys by the
 * [pathname] ONLY; the [search] string (query parameters / hash — filters, sort, anchors) is deliberately
 * excluded so those changes never trigger a re-fade, exactly as the web component documents. Exposed so a binding
 * layer that observes the whole location can hand the surface a correct, search-insensitive key.
 */
@Suppress("UnusedParameter")
fun routeTransitionKey(
    pathname: String,
    search: String,
): String = pathname

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). The `view.opened` event carries only the constant
 * surface [SLUG] — never the route, the page content or any user data — so observability can never leak which
 * screen a user navigated to. Kept free of Compose so it is unit-tested with a recording [Logger].
 */
object RouteTransitionDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = ROUTE_TRANSITION_SLUG

    /** The one-shot event emitted once when the surface opens. */
    const val EVENT_VIEW_OPENED: String = "view.opened"

    /** The structured-field key carrying the surface slug on the diagnostic. */
    const val FIELD_SURFACE: String = "surface"

    /**
     * Emits the one PII-safe `view.opened` diagnostic carrying only the surface [SLUG]. Call from the
     * composable's first-composition effect.
     */
    fun recordViewOpened(logger: Logger) {
        logger.info(EVENT_VIEW_OPENED, mapOf(FIELD_SURFACE to SLUG))
    }
}
